### PR TITLE
feat: Input prefix suffix

### PR DESCRIPTION
**Pull request by NathanZlion**
### Description

Related links, issue #, if available: n/a

### How has this been tested?

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pages/input/adornments.page.tsx [hunk 1]
```diff
@@ -0,0 +1,768 @@
...
+interface StateControlsProps {
+  disabled: boolean;
+  setDisabled: (v: boolean) => void;
+  readOnly: boolean;
+  setReadOnly: (v: boolean) => void;
+  invalid: boolean;
+  setInvalid: (v: boolean) => void;
+  warning: boolean;
+  setWarning: (v: boolean) => void;
+  compact: boolean;
```
Comment (jperals) [inline pages/input/adornments.page.tsx]: Compact density is already in the top nav controls (and RTL should ideally be as well at some point)

File: pages/input/adornments.page.tsx [hunk 2]
```diff
@@ -0,0 +1,768 @@
...
+          </ColumnLayout>
+        </div>
+      </Box>
+    </Container>
+  );
+}
+
+// ─── Main page ────────────────────────────────────────────────────────────────
+
+export default function AdornmentsPage() {
```
Comment (jperals) [inline pages/input/adornments.page.tsx]: For visual regression testing, a screenshot area will be needed.

My suggestion would be a separate permutations page, so that only the component to be tested is included in the screenshots and no other extraneous content which lead to unnecessary collateral diffs.

File: src/input/internal.tsx [hunk 3]
```diff
@@ -120,11 +125,29 @@ function InternalInput(
   __rightIcon = __rightIcon ?? searchProps.__rightIcon;
   __onRightIconClick = __onRightIconClick ?? searchProps.__onRightIconClick;
 
+  // Search inputs use built-in search and clear icons that would overlap adornments.
```
Comment (jperals) [inline src/input/internal.tsx:126]: Sounds like a good case for a dev warning?

File: src/input/internal.tsx [hunk 4]
```diff
@@ -120,11 +125,29 @@ function InternalInput(
...
+
   const formFieldContext = useFormFieldContext(rest);
   const { ariaLabelledby, ariaDescribedby, controlId, invalid, warning } = __inheritFormFieldProps
     ? formFieldContext
     : rest;
 
+  const hasAdornmentContent = (content: React.ReactNode) =>
+    content !== null && content !== undefined && typeof content !== 'boolean' && content !== '';
+  const hasPrefix = hasAdornmentContent(prefix);
+  const hasSuffix = hasAdornmentContent(suffix);
```
Comment (jperals) [inline src/input/internal.tsx]: Non blocking: Could this be just
```
const hasPrefix = !!prefix;
const hasSuffix = !!suffix;
```

I think the only possible difference would arise if someone sets `prefix={true}` or `suffix={true}`, but I would not expect that?

File: src/input/internal.tsx [hunk 5]
```diff
@@ -120,11 +125,29 @@ function InternalInput(
...
+  const hasAdornment = hasPrefix || hasSuffix;
```
Comment (jperals) [inline src/input/internal.tsx]: I find "adornment" not very clear, although I won't block on this because I can't think of perfect alternatives.

I think I would even prefer `hasPrefixOrSuffix`. Alternatively `hasAffix` ("affix" includes prefix and suffix) though not a common term.

File: src/input/internal.tsx [hunk 6]
```diff
@@ -211,10 +235,50 @@ function InternalInput(
...
+        )}
+      >
+        {mainInput}
+      </div>
+    </div>
+  ) : (
+    mainInput
+  );
+
+  const rightIcon = __rightIcon ? (
```
Comment (jperals) [inline src/input/internal.tsx]: Non-blocking because it doesn't technically break anything, but this terminology is not RTL-friendly and it is easy to name these elements like `suffixIcon`, and even clearer IMO.

Same with class names e.g, `input-icon-suffix`.

File: src/test-utils/dom/input/index.ts [hunk 7]
```diff
@@ -12,4 +12,18 @@ export default class InputWrapper extends BaseInputWrapper {
   findClearButton(): ElementWrapper | null {
     return this.find(`.${inputSelectors['input-button-right']}`);
   }
+
+  /**
+   * Returns the prefix adornment element, or null if no prefix is set.
```
Comment (jperals) [inline src/test-utils/dom/input/index.ts]: Again I wouldn't call this blocking but an adornment is purely decorational while this is not necessarily the case since it has a functional purpose in helping the user understanding how the input value will be used etc.

File: src/input/styles.scss [hunk 8]
```diff
@@ -224,3 +239,146 @@
...
+    ) {
+      padding-inline-start: 0;
+    }
+
+    &:focus-within {
+      box-shadow: var(#{custom-props.$styleBoxShadowFocus}, foundation.$box-shadow-focused-light-invalid);
+    }
+  }
+
+  &.input-adorned-container-disabled {
```
Comment (jperals) [inline src/input/styles.scss:309]: For some reason now the height of an input with prefix or suffix jumps from 20px to 22px when disabled or readOnly.

File: pages/input/prefix-suffix.page.tsx
Comment (gethinwebster) [inline pages/input/prefix-suffix.page.tsx:1]: suggest renaming prefix-suffix or similar to match feature naming

File: src/input/internal.tsx (hunk 4 above)
Comment (gethinwebster) [reply src/input/internal.tsx]: not sure why this was resolved? I think it's valid: the current code is overly complicated (and different to how we treat similar properties in other components)

File: src/test-utils/dom/input/index.ts [hunk 9]
```diff
@@ -10,6 +10,14 @@ export default class InputWrapper extends BaseInputWrapper {
   static rootSelector: string = inputSelectors.root;
 
   findClearButton(): ElementWrapper | null {
-    return this.find(`.${inputSelectors['input-button-right']}`);
+    return this.find(`.${inputSelectors['input-button-end']}`);
```
Comment (gethinwebster) [inline src/test-utils/dom/input/index.ts]: Ideally we should try not to change class names on existing test utils, as it causes backwards-incompatibility. As it's just a naming thing here, I suggest changing this one classname back.

File: src/input/styles.scss [hunk 10]
```diff
@@ -224,3 +239,184 @@
...
+.input-suffix {
+  display: flex;
+  align-items: center;
+  flex-shrink: 1;
+  max-inline-size: max(0px, calc((100% - 12rem) / 2));
+  min-inline-size: 0;
+  padding-block: styles.$control-padding-vertical;
+  padding-inline: styles.$control-padding-horizontal;
+  color: inherit;
+  user-select: text;
```
Comment (gethinwebster) [inline src/input/styles.scss]: do we need to specify these, I think they're the default anyway?

File: src/test-utils/dom/input/index.ts (hunk 9 above)
Comment (NathanZlion) [reply src/test-utils/dom/input/index.ts]: Odd edge-case to the !!hasPrefix is for a number 0 for instance. But I guess it's not a real usecase? it's falsy so doesn't render.

Comment (gethinwebster) [reply src/test-utils/dom/input/index.ts]: I think it's unlikely, and if it's needed it can be passed as string to make it truthy

**Review by gethinwebster — APPROVED** [review]
